### PR TITLE
Calendar interactor

### LINKED ISSUE #2792: Modify the InteractorFactory to introduce the appropriate Calendar methods
Ref #2754 for architectural details
Needs #2789 and #2790

We need to introduce an interactor so that the ViewModel layer doesn't consume the services themselves. This is a core part of the project and needs to be thoroughly tested.

### PR DESCRIPTION
**Pull request by juanlaube**
## What's this?

Adds a new interactor to fetch the calendar items for a given date. Items could be fetched from the calendar service or from existing time entries.

### Relationships

Closes #2792 

## Why do we want this?

This is what we will use to populate the calendar view.

## How is it done?

It's a new interactor that combines both observables from the calendar service and the time entries data source and sorts the items based on the start date.

### Side effects

None.

## :squid: Permissions

Whenever.

### CODE REVIEW

File: Toggl.Foundation/Calendar/CalendarItem.cs [hunk 1]
```diff
@@ -39,9 +39,12 @@ public CalendarItem(IThreadSafeTimeEntry timeEntry)
                 timeEntry.Start,
                 TimeSpan.FromSeconds(timeEntry.Duration.Value),
                 timeEntry.Description,
-                timeEntry.Project.Color,
+                timeEntry.Project?.Color ?? string.Empty,
```
Comment (willsb) [inline Toggl.Foundation/Calendar/CalendarItem.cs]: Use the `NoProject` color string instead

File: Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        private IObservable<IEnumerable<CalendarItem>> calendarItemsFromTimeEntries()
+            => timeEntriesDataSource.GetAll(timeEntry => timeEntry.Start > date.Date && timeEntry.Start < date.AddDays(1).Date && timeEntry.Duration != null)
+                .Select(convertTimeEntriesToCalendarItems);
+
+        private IObservable<IEnumerable<CalendarItem>> calendarItemsFromEvents()
+            => calendarService.GetEventsForDate(date);
+
+        private IEnumerable<CalendarItem> convertTimeEntriesToCalendarItems(IEnumerable<IThreadSafeTimeEntry> timeEntries)
+            => timeEntries.Select(timeEntry => new CalendarItem(timeEntry));
```
Comment (willsb) [inline Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs]: Why not `Select(CalendarItem.FromTimeEntry)`?

File: Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        public IObservable<IEnumerable<CalendarItem>> Execute()
+            => Observable.CombineLatest(
+                    calendarItemsFromTimeEntries(),
+                    calendarItemsFromEvents(),
+                    (timeEntries, events) => timeEntries.Concat(events))
+                .Select(orderByStartTime);
+
+        private IObservable<IEnumerable<CalendarItem>> calendarItemsFromTimeEntries()
+            => timeEntriesDataSource.GetAll(timeEntry => timeEntry.Start > date.Date && timeEntry.Start < date.AddDays(1).Date && timeEntry.Duration != null)
```
Comment (willsb) [inline Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs]: Shouldn't the comparisons be `>=` and `<=`?

File: Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            Ensure.Argument.IsNotNull(calendarService, nameof(calendarService));
+            Ensure.Argument.IsNotNull(date, nameof(date));
+
+            this.timeEntriesDataSource = timeEntriesDataSource;
+            this.calendarService = calendarService;
+            this.date = date;
+        }
+
+        public IObservable<IEnumerable<CalendarItem>> Execute()
+            => Observable.CombineLatest(
```
Comment (willsb) [inline Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs]: This looks weird at first, but I get it. Maybe something with defer would make sense, but I'll leave this as is for the time being

File: Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        private readonly DateTime date;
+
+        public GetCalendarItemsForDateInteractor(
+            ITimeEntriesSource timeEntriesDataSource,
+            ICalendarService calendarService,
+            DateTime date)
+        {
+            Ensure.Argument.IsNotNull(timeEntriesDataSource, nameof(timeEntriesDataSource));
+            Ensure.Argument.IsNotNull(calendarService, nameof(calendarService));
+            Ensure.Argument.IsNotNull(date, nameof(date));
```
Comment (arturs105) [inline Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs]: I don't think `date` can be null

File: Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs (hunk 3 above)
Comment (arturs105) [reply Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs]: I had the same question.

File: Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs (hunk 5 above)
Comment (juanlaube) [reply Toggl.Foundation/Interactors/Calendar/GetCalendarItemsForDateInteractor.cs]: but date can't be null, am I missing something? 🤔

**Pull request merged**